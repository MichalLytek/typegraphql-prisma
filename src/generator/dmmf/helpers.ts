export const modelAttributeRegex = /(@@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const fieldAttributeRegex = /(@TypeGraphQL\.)+([A-z])+(\()+(.+)+(\))+/;
export const attributeNameRegex = /(?:\.)+([A-Za-z])+(?:\()+/;
export const attributeArgsRegex = /(?:\()+([A-Za-z])+\:+(.+)+(?:\))+/;

export function parseDocumentationAttributes<TData extends object = object>(
  documentation: string | undefined,
  expectedAttributeName: string,
  expectedAttributeKind: "model" | "field",
): Partial<TData> {
  const attributeRegex =
    expectedAttributeKind === "model"
      ? modelAttributeRegex
      : fieldAttributeRegex;
  for (const documentation_ of documentation?.split("\\n") ?? []) {
    const attribute = documentation_?.match(attributeRegex)?.[0];
    const attributeName = attribute
      ?.match(attributeNameRegex)?.[0]
      ?.slice(1, -1);
    if (attributeName !== expectedAttributeName) {
      continue;
    }
    const rawAttributeArgs = attribute
      ?.match(attributeArgsRegex)?.[0]
      ?.slice(1, -1);
    const splitRawArgsArray = rawAttributeArgs?.split(",").map(it => it.trim());
    const parsedAttributeArgs =
      splitRawArgsArray &&
      (Object.fromEntries(
        splitRawArgsArray.map(it => {
          const [key, value] = it.split(": ");
          return [key, JSON.parse(value)];
        }),
      ) as Partial<TData>);
    if (!parsedAttributeArgs) {
      continue;
    }
    return parsedAttributeArgs;
  }
  return {};
}
