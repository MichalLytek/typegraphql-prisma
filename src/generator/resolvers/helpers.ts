import { OptionalKind, MethodDeclarationStructure, Writers } from "ts-morph";

import { DmmfDocument } from "../dmmf/dmmf-document";
import { DMMF } from "../dmmf/types";
import { GeneratorOptions } from "../options";

export function generateCrudResolverClassMethodDeclaration(
  action: DMMF.Action,
  mapping: DMMF.ModelMapping,
  dmmfDocument: DmmfDocument,
  generatorOptions: GeneratorOptions,
): OptionalKind<MethodDeclarationStructure> {
  return {
    name: action.name,
    isAsync: true,
    returnType: `Promise<${action.returnTSType}>`,
    decorators: [
      {
        name: `TypeGraphQL.${action.operation}`,
        arguments: [
          `_returns => ${action.typeGraphQLType}`,
          Writers.object({
            nullable: `${!action.method.isRequired}`,
          }),
        ],
      },
    ],
    parameters: [
      {
        name: "ctx",
        // TODO: import custom `ContextType`
        type: "any",
        decorators: [{ name: "TypeGraphQL.Ctx", arguments: [] }],
      },
      {
        name: "info",
        type: "GraphQLResolveInfo",
        decorators: [{ name: "TypeGraphQL.Info", arguments: [] }],
      },
      ...(!action.argsTypeName
        ? []
        : [
            {
              name: "args",
              type: action.argsTypeName,
              decorators: [
                {
                  name: "TypeGraphQL.Args",
                  arguments: generatorOptions.emitRedundantDecoratorArgs
                    ? [`_returns => ${action.argsTypeName}`]
                    : [],
                },
              ],
            },
          ]),
    ],
    statements:
      action.kind === DMMF.ModelAction.aggregate
        ? [
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...transformInfoIntoPrismaArgs(info),
            });`,
          ]
        : action.kind === DMMF.ModelAction.groupBy
        ? [
            /* ts */ ` const { _count, _avg, _sum, _min, _max } = transformInfoIntoPrismaArgs(info);`,
            /* ts */ ` return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...Object.fromEntries(
                Object.entries({ _count, _avg, _sum, _min, _max }).filter(([_, v]) => v != null)
              ),
            });`,
          ]
        : [
            /* ts */ ` const { _count } = transformInfoIntoPrismaArgs(info);
            return getPrismaFromContext(ctx).${mapping.collectionName}.${action.prismaMethod}({
              ...args,
              ...(_count && transformCountFieldIntoSelectRelationsCount(_count)),
            });`,
          ],
  };
}
