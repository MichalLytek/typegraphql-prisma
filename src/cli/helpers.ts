import { GeneratorOptions } from "@prisma/generator-helper";
import { DMMF } from "../generator/types";

export async function loadFileCallback(
  filePath: string | undefined,
):
  | undefined
  | ((
      actionName: DMMF.ModelAction,
      typeName: string,
      options: GeneratorOptions,
    ) => string) {
  if (!filePath) {
    return;
  }

  const callback = await import(filePath);
  if (typeof callback !== "function") {
    throw new Error(`Invalid callback, callback must be a function`);
  }

  return callback;
}

export function parseStringBoolean(stringBoolean: string | undefined) {
  return stringBoolean ? stringBoolean === "true" : undefined;
}

export function parseStringArray<TAllowedValue extends string>(
  stringArray: string | undefined,
  optionPropertyName: string,
  allowedValues?: TAllowedValue[],
): TAllowedValue[] | undefined {
  if (!stringArray) {
    return undefined;
  }
  const parsedArray = stringArray.split(",").map(it => it.trim());
  if (allowedValues) {
    for (const option of parsedArray) {
      if (!allowedValues.includes(option as any)) {
        throw new Error(
          `Invalid "${optionPropertyName}" option value "${option}" provided for TypeGraphQL generator.`,
        );
      }
    }
  }
  return parsedArray as TAllowedValue[];
}
