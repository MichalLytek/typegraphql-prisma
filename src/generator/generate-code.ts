import { DMMF as PrismaDMMF } from "@prisma/client/runtime";
import { Project, ScriptTarget, ModuleKind, CompilerOptions } from "ts-morph";
import path from "path";

import { noop, toUnixPath } from "./helpers";
import generateEnumFromDef from "./enum";
import generateObjectTypeClassFromModel from "./model-type-class";
import generateRelationsResolverClassesFromModel from "./resolvers/relations";
import {
  generateOutputTypeClassFromType,
  generateInputTypeClassFromType,
} from "./type-class";
import generateCrudResolverClassFromMapping from "./resolvers/full-crud";
import {
  resolversFolderName,
  relationsResolversFolderName,
  crudResolversFolderName,
  inputsFolderName,
  outputsFolderName,
  enumsFolderName,
  modelsFolderName,
  argsFolderName,
} from "./config";
import {
  generateResolversBarrelFile,
  generateInputsBarrelFile,
  generateOutputsBarrelFile,
  generateIndexFile,
  generateModelsBarrelFile,
  generateEnumsBarrelFile,
  generateArgsBarrelFile,
  generateArgsIndexFile,
  generateResolversIndexFile,
  generateResolversActionsBarrelFile,
} from "./imports";
import {
  InternalGeneratorOptions,
  ExternalGeneratorOptions,
  GeneratorOptions,
} from "./options";
import { DmmfDocument } from "./dmmf/dmmf-document";
import generateArgsTypeClassFromArgs from "./args-class";
import generateActionResolverClass from "./resolvers/separate-action";
import { ensureInstalledCorrectPrismaPackage } from "../utils/prisma-version";
import { GenerateMappingData } from "./types";
import { generateEnhanceMap } from "./generate-enhance";
import { generateCustomScalars } from "./generate-scalars";
import { generateHelpersFile } from "./generate-helpers";
import { DMMF } from "./dmmf/types";
import { getBlocksToEmit } from "./emit-block";

const baseCompilerOptions: CompilerOptions = {
  target: ScriptTarget.ES2019,
  module: ModuleKind.CommonJS,
  emitDecoratorMetadata: true,
  experimentalDecorators: true,
  esModuleInterop: true,
};

export default async function generateCode(
  dmmf: PrismaDMMF.Document,
  baseOptions: InternalGeneratorOptions & ExternalGeneratorOptions,
  log: (msg: string) => void = noop,
) {
  ensureInstalledCorrectPrismaPackage();

  const options: GeneratorOptions = {
    ...baseOptions,
    blocksToEmit: getBlocksToEmit(baseOptions.emitOnly),
    contextPrismaKey: baseOptions.contextPrismaKey ?? "prisma",
    relativePrismaOutputPath: toUnixPath(
      path.relative(baseOptions.outputDirPath, baseOptions.prismaClientPath),
    ),
    absolutePrismaOutputPath:
      !baseOptions.customPrismaImportPath &&
      baseOptions.prismaClientPath.includes("node_modules")
        ? "@prisma/client"
        : undefined,
  };

  const baseDirPath = options.outputDirPath;
  const emitTranspiledCode =
    options.emitTranspiledCode ??
    options.outputDirPath.includes("node_modules");
  const project = new Project({
    compilerOptions: {
      ...baseCompilerOptions,
      ...(emitTranspiledCode && {
        declaration: true,
        importHelpers: true,
      }),
    },
  });
  const resolversDirPath = path.resolve(baseDirPath, resolversFolderName);

  log("Transforming dmmfDocument...");
  const dmmfDocument = new DmmfDocument(dmmf, options);

  if (dmmfDocument.shouldGenerateBlock("enums")) {
    log("Generating enums...");
    const datamodelEnumNames = dmmfDocument.datamodel.enums.map(
      enumDef => enumDef.typeName,
    );
    dmmfDocument.datamodel.enums.forEach(enumDef =>
      generateEnumFromDef(project, baseDirPath, enumDef),
    );
    dmmfDocument.schema.enums
      // skip enums from datamodel
      .filter(enumDef => !datamodelEnumNames.includes(enumDef.typeName))
      .forEach(enumDef => generateEnumFromDef(project, baseDirPath, enumDef));
    const emittedEnumNames = [
      ...new Set([
        ...dmmfDocument.schema.enums.map(it => it.typeName),
        ...dmmfDocument.datamodel.enums.map(it => it.typeName),
      ]),
    ];
    const enumsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(baseDirPath, enumsFolderName, "index.ts"),
      undefined,
      { overwrite: true },
    );
    generateEnumsBarrelFile(enumsBarrelExportSourceFile, emittedEnumNames);
  }

  if (dmmfDocument.shouldGenerateBlock("models")) {
    log("Generating models...");
    dmmfDocument.datamodel.models.forEach(model => {
      const modelOutputType = dmmfDocument.schema.outputTypes.find(
        type => type.name === model.name,
      )!;
      return generateObjectTypeClassFromModel(
        project,
        baseDirPath,
        model,
        modelOutputType,
        dmmfDocument,
      );
    });
    const modelsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(baseDirPath, modelsFolderName, "index.ts"),
      undefined,
      { overwrite: true },
    );
    generateModelsBarrelFile(
      modelsBarrelExportSourceFile,
      dmmfDocument.datamodel.models.map(it => it.typeName),
    );
  }

  let outputTypesToGenerate: DMMF.OutputType[] = [];
  if (dmmfDocument.shouldGenerateBlock("outputs")) {
    log("Generating output types...");
    const rootTypes = dmmfDocument.schema.outputTypes.filter(type =>
      ["Query", "Mutation"].includes(type.name),
    );
    const modelNames = dmmfDocument.datamodel.models.map(model => model.name);
    outputTypesToGenerate = dmmfDocument.schema.outputTypes.filter(
      // skip generating models and root resolvers
      type => !modelNames.includes(type.name) && !rootTypes.includes(type),
    );
    const outputTypesFieldsArgsToGenerate = outputTypesToGenerate
      .map(it => it.fields)
      .reduce((a, b) => a.concat(b), [])
      .filter(it => it.argsTypeName);
    outputTypesToGenerate.forEach(type =>
      generateOutputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
      ),
    );
    const outputsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        outputsFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateOutputsBarrelFile(
      outputsBarrelExportSourceFile,
      outputTypesToGenerate.map(it => it.typeName),
      outputTypesFieldsArgsToGenerate.length > 0,
    );

    if (outputTypesFieldsArgsToGenerate.length > 0) {
      log("Generating output types args...");
      outputTypesFieldsArgsToGenerate.forEach(async field => {
        generateArgsTypeClassFromArgs(
          project,
          path.resolve(resolversDirPath, outputsFolderName),
          field.args,
          field.argsTypeName!,
          dmmfDocument,
          2,
        );
      });
      const outputsArgsBarrelExportSourceFile = project.createSourceFile(
        path.resolve(
          baseDirPath,
          resolversFolderName,
          outputsFolderName,
          argsFolderName,
          "index.ts",
        ),
        undefined,
        { overwrite: true },
      );
      generateArgsBarrelFile(
        outputsArgsBarrelExportSourceFile,
        outputTypesFieldsArgsToGenerate.map(it => it.argsTypeName!),
      );
    }
  }

  if (dmmfDocument.shouldGenerateBlock("inputs")) {
    log("Generating input types...");
    dmmfDocument.schema.inputTypes.forEach(type =>
      generateInputTypeClassFromType(
        project,
        resolversDirPath,
        type,
        dmmfDocument,
        options,
      ),
    );
    const inputsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        inputsFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateInputsBarrelFile(
      inputsBarrelExportSourceFile,
      dmmfDocument.schema.inputTypes.map(it => it.typeName),
    );
  }

  if (
    dmmfDocument.relationModels.length > 0 &&
    dmmfDocument.shouldGenerateBlock("relationResolvers")
  ) {
    log("Generating relation resolvers...");
    dmmfDocument.relationModels.forEach(relationModel =>
      generateRelationsResolverClassesFromModel(
        project,
        baseDirPath,
        dmmfDocument,
        relationModel,
        options,
      ),
    );
    const relationResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "resolvers.index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      relationResolversBarrelExportSourceFile,
      dmmfDocument.relationModels.map<GenerateMappingData>(relationModel => ({
        resolverName: relationModel.resolverName,
        modelName: relationModel.model.typeName,
      })),
    );

    log("Generating relation resolver args...");
    dmmfDocument.relationModels.forEach(async relationModelData => {
      const resolverDirPath = path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        relationModelData.model.typeName,
      );
      relationModelData.relationFields
        .filter(field => field.argsTypeName)
        .forEach(async field => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            field.outputTypeField.args,
            field.argsTypeName!,
            dmmfDocument,
          );
        });
      const argTypeNames = relationModelData.relationFields
        .filter(it => it.argsTypeName !== undefined)
        .map(it => it.argsTypeName!);

      if (argTypeNames.length) {
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, "index.ts"),
          undefined,
          { overwrite: true },
        );
        generateArgsBarrelFile(barrelExportSourceFile, argTypeNames);
      }
    });

    const relationModelsWithArgs = dmmfDocument.relationModels.filter(
      relationModelData =>
        relationModelData.relationFields.some(
          it => it.argsTypeName !== undefined,
        ),
    );
    if (relationModelsWithArgs.length > 0) {
      const relationResolversArgsIndexSourceFile = project.createSourceFile(
        path.resolve(
          baseDirPath,
          resolversFolderName,
          relationsResolversFolderName,
          "args.index.ts",
        ),
        undefined,
        { overwrite: true },
      );
      generateArgsIndexFile(
        relationResolversArgsIndexSourceFile,
        relationModelsWithArgs.map(
          relationModelData => relationModelData.model.typeName,
        ),
      );
    }
    const relationResolversIndexSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        relationsResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversIndexFile(
      relationResolversIndexSourceFile,
      "relations",
      relationModelsWithArgs.length > 0,
    );
  }

  if (dmmfDocument.shouldGenerateBlock("crudResolvers")) {
    log("Generating crud resolvers...");
    dmmfDocument.modelMappings.forEach(async mapping => {
      const model = dmmfDocument.datamodel.models.find(
        model => model.name === mapping.modelName,
      )!;
      generateCrudResolverClassFromMapping(
        project,
        baseDirPath,
        mapping,
        model,
        dmmfDocument,
        options,
      );
      mapping.actions.forEach(async action => {
        const model = dmmfDocument.datamodel.models.find(
          model => model.name === mapping.modelName,
        )!;
        generateActionResolverClass(
          project,
          baseDirPath,
          model,
          action,
          mapping,
          dmmfDocument,
          options,
        );
      });
    });
    const generateMappingData =
      dmmfDocument.modelMappings.map<GenerateMappingData>(mapping => {
        const model = dmmfDocument.datamodel.models.find(
          model => model.name === mapping.modelName,
        )!;
        return {
          modelName: model.typeName,
          resolverName: mapping.resolverName,
          actionResolverNames: mapping.actions.map(it => it.actionResolverName),
        };
      });
    const crudResolversBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        "resolvers-crud.index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversBarrelFile(
      crudResolversBarrelExportSourceFile,
      generateMappingData,
    );
    const crudResolversActionsBarrelExportSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        "resolvers-actions.index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversActionsBarrelFile(
      crudResolversActionsBarrelExportSourceFile,
      generateMappingData,
    );
    const crudResolversIndexSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        "index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateResolversIndexFile(crudResolversIndexSourceFile, "crud", true);

    log("Generating crud resolvers args...");
    dmmfDocument.modelMappings.forEach(async mapping => {
      const actionsWithArgs = mapping.actions.filter(
        it => it.argsTypeName !== undefined,
      );

      if (actionsWithArgs.length) {
        const model = dmmfDocument.datamodel.models.find(
          model => model.name === mapping.modelName,
        )!;
        const resolverDirPath = path.resolve(
          baseDirPath,
          resolversFolderName,
          crudResolversFolderName,
          model.typeName,
        );
        actionsWithArgs.forEach(async action => {
          generateArgsTypeClassFromArgs(
            project,
            resolverDirPath,
            action.method.args,
            action.argsTypeName!,
            dmmfDocument,
          );
        });
        const barrelExportSourceFile = project.createSourceFile(
          path.resolve(resolverDirPath, argsFolderName, "index.ts"),
          undefined,
          { overwrite: true },
        );
        generateArgsBarrelFile(
          barrelExportSourceFile,
          actionsWithArgs.map(it => it.argsTypeName!),
        );
      }
    });
    const crudResolversArgsIndexSourceFile = project.createSourceFile(
      path.resolve(
        baseDirPath,
        resolversFolderName,
        crudResolversFolderName,
        "args.index.ts",
      ),
      undefined,
      { overwrite: true },
    );
    generateArgsIndexFile(
      crudResolversArgsIndexSourceFile,
      dmmfDocument.modelMappings
        .filter(mapping =>
          mapping.actions.some(it => it.argsTypeName !== undefined),
        )
        .map(mapping => mapping.modelTypeName),
    );
  }

  log("Generate enhance map");
  const enhanceSourceFile = project.createSourceFile(
    baseDirPath + "/enhance.ts",
    undefined,
    { overwrite: true },
  );
  generateEnhanceMap(
    enhanceSourceFile,
    dmmfDocument,
    dmmfDocument.modelMappings,
    dmmfDocument.relationModels,
    dmmfDocument.datamodel.models,
    dmmfDocument.schema.inputTypes,
    outputTypesToGenerate,
  );

  log("Generate custom scalars");
  const scalarsSourceFile = project.createSourceFile(
    baseDirPath + "/scalars.ts",
    undefined,
    { overwrite: true },
  );
  generateCustomScalars(scalarsSourceFile, dmmfDocument.options);

  log("Generate custom helpers");
  const helpersSourceFile = project.createSourceFile(
    baseDirPath + "/helpers.ts",
    undefined,
    { overwrite: true },
  );
  generateHelpersFile(helpersSourceFile, dmmfDocument.options);

  log("Generating index file");
  const indexSourceFile = project.createSourceFile(
    baseDirPath + "/index.ts",
    undefined,
    { overwrite: true },
  );
  generateIndexFile(
    indexSourceFile,
    dmmfDocument.relationModels.length > 0,
    dmmfDocument.options.blocksToEmit,
  );

  log("Emitting generated code files");
  if (emitTranspiledCode) {
    await project.emit();
  } else {
    for (const file of project.getSourceFiles()) {
      file.formatText({ indentSize: 2 });
    }
    await project.save();
  }
}
