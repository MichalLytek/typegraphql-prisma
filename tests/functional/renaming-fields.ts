import "reflect-metadata";
import { promises as fs } from "fs";
import { buildSchema } from "type-graphql";
import { graphql } from "graphql";

import generateArtifactsDirPath from "../helpers/artifacts-dir";
import { generateCodeFromSchema } from "../helpers/generate-code";

describe("crud resolvers execution", () => {
  let outputDirPath: string;

  beforeEach(async () => {
    outputDirPath = generateArtifactsDirPath("renaming-fields");
    await fs.mkdir(outputDirPath, { recursive: true });
  });

  it("should properly resolve aliased field values from prisma model values", async () => {
    const prismaSchema = /* prisma */ `
      model User {
        id           Int       @id @default(autoincrement())
        dateOfBirth  DateTime
        /// @TypeGraphQL.field(name: "firstName")
        name         String
      }
    `;
    await generateCodeFromSchema(prismaSchema, { outputDirPath });
    const { UserCrudResolver } = require(
      outputDirPath + "/resolvers/crud/User/UserCrudResolver.ts",
    );
    const graphQLSchema = await buildSchema({
      resolvers: [UserCrudResolver],
      validate: false,
    });
    const document = /* graphql */ `
      query {
        user(where: { id: 1 }) {
          id
          dateOfBirth
          firstName
        }
      }
    `;
    const prismaMock = {
      user: {
        findUnique: jest.fn().mockResolvedValue({
          id: 1,
          dateOfBirth: new Date("2019-12-31T14:16:02.572Z"),
          name: "John",
        }),
      },
    };

    const { data, errors } = await graphql({
      schema: graphQLSchema,
      source: document,
      contextValue: { prisma: prismaMock },
    });

    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot("user mocked response");
  });

  it("should properly map aliased input field values to prisma input values", async () => {
    const prismaSchema = /* prisma */ `
      model User {
        id           Int       @id @default(autoincrement())
        dateOfBirth  DateTime
        /// @TypeGraphQL.field(name: "firstName")
        name         String
      }
    `;
    await generateCodeFromSchema(prismaSchema, { outputDirPath });
    const { UserCrudResolver } = require(
      outputDirPath + "/resolvers/crud/User/UserCrudResolver.ts",
    );
    const graphQLSchema = await buildSchema({
      resolvers: [UserCrudResolver],
      validate: false,
    });
    const document = /* graphql */ `
      query {
        users(where: { firstName: { equals: "John" }}) {
          id
        }
      }
    `;
    const prismaMock = {
      user: {
        findMany: jest.fn().mockResolvedValue([{ id: 1 }]),
      },
    };

    const { errors } = await graphql({
      schema: graphQLSchema,
      source: document,
      contextValue: { prisma: prismaMock },
    });

    expect(errors).toBeUndefined();
    expect(prismaMock.user.findMany.mock.calls).toMatchSnapshot(
      "findManyUser call args",
    );
  });
});
