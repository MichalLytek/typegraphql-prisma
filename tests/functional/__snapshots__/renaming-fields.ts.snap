// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`crud resolvers execution should properly map aliased input field values to prisma input values: findManyUser call args 1`] = `
Array [
  Array [
    Object {
      "where": UserWhereInput {
        "name": StringFilter {
          "equals": "John",
        },
      },
    },
  ],
]
`;

exports[`crud resolvers execution should properly resolve aliased field values from prisma model values: user mocked response 1`] = `
Object {
  "user": Object {
    "dateOfBirth": "2019-12-31T14:16:02.572Z",
    "firstName": "John",
    "id": 1,
  },
}
`;
